#
# Provide optional sftp idle session timeout for ZFSSA only.  
# Not documented for Solaris.
#
# The sftp-server changes are rejected by upstream in favor of a longer-term,
# more-comprehensive idle session timeout that is in development.  So the
# upstream fix, when it is published, will supplant this temporary fix
# as soon as is feasible within the ZFSSA release cycle. The requirement
# is tracked by enhancement bz2718.
#
# The interface to sftp MUST NOT BE CHANGED here without coordinating
# with ZFSSA.
#
--- old/sftp-server.c
+++ new/sftp-server.c
@@ -36,6 +36,7 @@
 #include <time.h>
 #include <unistd.h>
 #include <stdarg.h>
+#include <signal.h>
 
 #include "atomicio.h"
 #include "xmalloc.h"
@@ -95,6 +96,9 @@ struct Stat {
 	Attrib attrib;
 };
 
+/* sftp idle timeout */
+static volatile sig_atomic_t g_timed_out = 0;
+
 /* Packet handlers */
 static void process_open(uint32_t id);
 static void process_close(uint32_t id);
@@ -2287,7 +2291,10 @@ sftp_server_cleanup_exit(int i)
 		    pw->pw_name, client_addr);
 	}
 #ifdef SFTP_AUDIT
-	audit_sftp_session_stop((i != 0) ? ADT_FAIL_VALUE_PROGRAM : 0);
+	if (i == 3)
+	    audit_sftp_session_stop(ADT_FAIL_VALUE_IDLE_TIMEOUT);
+	else
+	    audit_sftp_session_stop((i != 0) ? ADT_FAIL_VALUE_PROGRAM : 0);
 #endif
 	_exit(i);
 }
@@ -2306,10 +2313,17 @@ sftp_server_usage(void)
 	exit(1);
 }
 
+static void
+sftp_alarm_handler(int sig)
+{
+	g_timed_out = 1;
+}
+
 int
 sftp_server_main(int argc, char **argv, struct passwd *user_pw)
 {
 	int i, r, in, out, ch, skipargs = 0, log_stderr = 0;
+	int timeout = 0;
 	ssize_t len, olen;
 	SyslogFacility log_facility = SYSLOG_FACILITY_AUTH;
 	char *cp, *homedir = NULL, uidstr[32], buf[4*4096];
@@ -2324,7 +2338,7 @@ sftp_server_main(int argc, char **argv, struct passwd *user_pw)
 	pw = pwcopy(user_pw);
 
 	while (!skipargs && (ch = getopt(argc, argv,
-	    "d:f:l:P:p:Q:u:cehR")) != -1) {
+	    "d:f:l:P:p:Q:t:u:cehR")) != -1) {
 		switch (ch) {
 		case 'Q':
 			if (strcasecmp(optarg, "requests") != 0) {
@@ -2386,6 +2400,20 @@ sftp_server_main(int argc, char **argv, struct passwd *user_pw)
 				fatal("Invalid umask \"%s\"", optarg);
 			(void)umask((mode_t)mask);
 			break;
+		case 't':
+			timeout = atoi(optarg);
+			debug("SFTP: optarg = %s\n", optarg);
+			debug("SFTP: setting timeout to %d\n", timeout);
+			if (timeout < 0) {
+				/*
+				 * -t0 is redundant, but will be used
+				 * by ZFSSA BUI/CLI
+				 */
+				error("Invalid timeout: \"%s\". Ignored.",
+				    optarg);
+				timeout = 0;
+			}
+			break;
 		case 'h':
 		default:
 			sftp_server_usage();
@@ -2447,6 +2475,7 @@ sftp_server_main(int argc, char **argv, struct passwd *user_pw)
 		}
 	}
 
+	signal(SIGALRM, sftp_alarm_handler);
 	for (;;) {
 		struct pollfd pfd[2];
 
@@ -2473,13 +2502,23 @@ sftp_server_main(int argc, char **argv, struct passwd *user_pw)
 			pfd[1].events = POLLOUT;
 		}
 
+		if (timeout > 0)
+			alarm(timeout);
+
 		if (poll(pfd, 2, -1) == -1) {
+			if (g_timed_out){
+				debug("SFTP idle timeout, closing session.");
+				sftp_server_cleanup_exit(3);
+			}
 			if (errno == EINTR)
 				continue;
 			error("poll: %s", strerror(errno));
 			sftp_server_cleanup_exit(2);
 		}
 
+		if (timeout > 0)
+			alarm(0);
+
 		/* copy stdin to iqueue */
 		if (pfd[0].revents & (POLLIN|POLLHUP)) {
 			len = read(in, buf, sizeof buf);
